fl4health/clients/fenda_client.py

```diff
@@ -1,12 +1,220 @@
-from flwr.common.typing import Config
+import copy
+from pathlib import Path
+from typing import Dict, List, Optional, Sequence, Tuple
 
+import torch
+from flwr.common.typing import Config, NDArrays
+
+from fl4health.checkpointing.checkpointer import TorchCheckpointer
 from fl4health.clients.basic_client import BasicClient
 from fl4health.model_bases.fenda_base import FendaModel
 from fl4health.parameter_exchange.layer_exchanger import FixedLayerExchanger
 from fl4health.parameter_exchange.parameter_exchanger_base import ParameterExchanger
+from fl4health.utils.losses import Losses, LossMeterType
+from fl4health.utils.metrics import Metric, MetricMeterType
 
 
 class FendaClient(BasicClient):
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        metric_meter_type: MetricMeterType = MetricMeterType.AVERAGE,
+        checkpointer: Optional[TorchCheckpointer] = None,
+        temperature: float = 0.5,
+        perFCL_loss_weights: List[float] = [0.0, 0.0],
+        cos_sim_loss_weight: float = 0.0,
+        contrastive_loss_weight: float = 0.0,
+    ) -> None:
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            metric_meter_type=metric_meter_type,
+            checkpointer=checkpointer,
+        )
+        self.perFCL_loss_weights = perFCL_loss_weights
+        self.cos_sim_loss_weight = cos_sim_loss_weight
+        self.contrastive_loss_weight = contrastive_loss_weight
+        self.cos_sim = torch.nn.CosineSimilarity(dim=-1)
+        self.ce_criterion = torch.nn.CrossEntropyLoss().to(self.device)
+        self.temperature: float = temperature
+
+        # Need to save previous local module, global module and aggregated global module at each communication round
+        # to compute contrastive loss.
+        self.old_local_module: torch.nn.Module
+        self.old_global_module: torch.nn.Module
+        self.aggregated_global_module: torch.nn.Module
+
     def get_parameter_exchanger(self, config: Config) -> ParameterExchanger:
         assert isinstance(self.model, FendaModel)
         return FixedLayerExchanger(self.model.layers_to_exchange())
+
+    def predict(self, input: torch.Tensor) -> Dict[str, torch.Tensor]:
+
+        preds = self.model(input)
+
+        preds["old_local_features"] = self.old_local_module.forward(input).view(len(preds["local_features"]), -1)
+        preds["old_global_features"] = self.old_global_module.forward(input).view(len(preds["global_features"]), -1)
```

Comment:
I think, but I'm not sure, that `reshape` is a more general tensor transformation than `view`. That is, reshape will perform a view if possible and return a copy if not. I'm not 100% sure though.

I believe that we're making an assumption about how the features are structured here. That is, I think we're assuming they are "batch first"? If that's true we should at least leave a comment.

Reply:
It also looks like we're already performing this operation in the forward of the `fenda_base`. Do we want to assume that it has already been done in the forward and eliminate the operation here, or not perform the flattening in the forward and perform it here?

Reply:
Oh no for global module and local module it is not done yet. In fenda model I do the reshape (`.view`) in the head module, not in each local and global modules separately. That's why we are sure of preds["local_features"] and preds["global_features"] dimensionality and here we reshape the output of each module accordingly.

Reply:
Yes that is true, I rechecked the difference between `reshape` and `view`. Although I don't think `view` would be problematic, I will change it to `reshape` to be more generalized. Thanks for mentioning.
Also yes we assume that batch should be the first dimension. I will add the comment.